### PR TITLE
When grossman is slow, retry 3 times before yielding an error

### PR DESCRIPTION
**Pull request by michaelachrisco**
Fixes #71 on Payroll

### CODE REVIEW

**Comment by michaelachrisco** [comment]
@josephbridgwaterrowe Please review.

File: lib/gman_client.rb [hunk 1]
```diff
@@ -14,40 +16,47 @@ def initialize(options = {})
       @client_id = options[:client_id] || ENV['CLIENT_ID']
       @client_secret = options[:client_secret] || ENV['CLIENT_SECRET']
     end
-    # class << self
 
     def drivers
-      response = request.api.v1.drivers.get
+      response = attempt_thrince { request.api.v1.drivers.get }
```
Comment (josephbridgwaterrowe) [inline lib/gman_client.rb]: `thrice`?

**Comment by josephbridgwaterrowe** [comment]
Looks good, rebase and correct spelling?

**Comment by michaelachrisco** [comment]
Thanks. I blame the end of the day.

**Comment by josephbridgwaterrowe** [comment]
I'm going to merge on Monday when I'm a bit sharper and we can discuss retries... I would like to give this a whirl before we put into production.

**Comment by michaelachrisco** [comment]
Absolutely. I am trying it out now in a separate branch in payroll.

**Comment by michaelachrisco** [comment]
We can also use the retry pattern: http://blog.mirthlab.com/2012/05/25/cleanly-retrying-blocks-of-code-after-an-exception-in-ruby/
if we want to make sure no real magic is happening within the code. Although the retries gem is super simple https://github.com/ooyala/retries/blob/master/lib/retries.rb. Its your call.

File: lib/gman_client.rb [hunk 2]
```diff
@@ -69,9 +78,16 @@ def retrieve_token
...
+    private
+
     def convert_payload(response)
-      response.payload.first.nil? ? [response.to_h] : response.map(&:to_h)
+      response.map do |hash_request|
+        Hashie.symbolize_keys!(hash_request.to_h)
+      end
+    end
+
+    def attempt_thrice
```
Comment (josephbridgwaterrowe) [inline lib/gman_client.rb]: How about renaming to `attempt(retry_count)` and passing `retry_count` to `:max_tries`?

Comment (michaelachrisco) [inline lib/gman_client.rb]: Ill do `retry_count` just in case the user wants to override the count.

File: lib/gman_client.rb [hunk 3]
```diff
@@ -69,9 +79,16 @@ def retrieve_token
...
+    private
+
     def convert_payload(response)
-      response.payload.first.nil? ? [response.to_h] : response.map(&:to_h)
+      response.map do |hash_request|
+        Hashie.symbolize_keys!(hash_request.to_h)
+      end
+    end
+
+    def retry_count
```
Comment (josephbridgwaterrowe) [inline lib/gman_client.rb]: I think this looks wanky now.

Comment (michaelachrisco) [inline lib/gman_client.rb]: Don't like `max_retry_count`? I could remove and let it be for now...

File: lib/gman_client.rb [hunk 4]
```diff
@@ -69,9 +78,16 @@ def retrieve_token
...
+    private
+
     def convert_payload(response)
-      response.payload.first.nil? ? [response.to_h] : response.map(&:to_h)
+      response.map do |hash_request|
+        Hashie.symbolize_keys!(hash_request.to_h)
+      end
+    end
+
+    def attempt(retry_count)
```
Comment (josephbridgwaterrowe) [inline lib/gman_client.rb]: I prefer this, how do you feel about it?

I didn't mean to steam roll over the existing implementation (which in reflection I can see is how may previous comment would be taken), so I'm sorry about that.

I'm quite open to a discussion about the other approach if you want. :smile:

**Comment by michaelachrisco** [comment]
I am fine with it. If we want to add anything additional, we can do so when appropriate. Rebase?

File: lib/gman_client.rb (hunk 4 above)
Comment (josephbridgwaterrowe) [inline lib/gman_client.rb]: @michaelachrisco, do you have any thoughts on this matter?

**Comment by josephbridgwaterrowe** [comment]
Yes, please rebase.

**Comment by josephbridgwaterrowe** [comment]
Thanks!

**Comment by michaelachrisco** [comment]
Cool. Thanks.

**Pull request merged**